WordPress/gutenberg
# Fix constrained tabbing failures with  Safari and Firefox

Fixes https://github.com/WordPress/gutenberg/issues/46041
Fixes https://github.com/WordPress/gutenberg/issues/41500

## What?

The refactoring of `useConstrainedTabbing` in https://github.com/WordPress/gutenberg/pull/34836 introduced a few regressions and unexpected behaviors. Some of these regressions have been fixed already, see for example https://github.com/WordPress/gutenberg/issues/42652, https://github.com/WordPress/gutenberg/issues/45903, ,and https://github.com/WordPress/gutenberg/pull/42653. 

This PR aims to fix one remaining regression with Safari and some unexpected behavior with Firefox.

## TL;DR
The regression and unexpected behaviors surfaced only when attempting to run the existing [editor/various/a11y.spec.js](https://github.com/WordPress/gutenberg/blob/98ce5c53c9492764aeff88852a4073fb32e784ac/test/e2e/specs/editor/various/a11y.spec.js) test with all browsers in https://github.com/WordPress/gutenberg/pull/46038. The outcome was 1 failure with  webkit (Safari) and 2 failures with Firefox.

That proved the native browsers behavior differs in some cases. Therefore, the current implementation of `useConstrainedTabbing`, which relies on native browser behavior, doesn't appear to be ideal in the first place.

I'd like to note that when it comes to some accessibility features, it is very, very important to run E2E tests with all ajor browsers. In fact, it turned out browser behavior differs in at least two cases:
- Safari: when clicking somewhere after the last tabbable within a modal dialog and then pressing Tab, the modal dialog closes unexpectedly. Instead, wIth Chrome and Firefox the tab sequence restarts from the previous tababble. This seems related to the so called [Sequential focus navigation starting point](https://chromestatus.com/feature/5671747802103808), which is a browser feature that is apparently implemented differently in Safari.
- Firefox: it's the only major browser that natively makes scrollable elements focusable. The `useConstrainedTabbing` current implementation allows to tab to scrollable elements but it doesn't take them into account as 'tabbable' elements, leading to unexpected behavior.

I'd also like to kindly note that it appears the refactoring in https://github.com/WordPress/gutenberg/pull/34836 didn[t introduce any new test for the specific use case it was meant to address, see https://github.com/WordPress/gutenberg/issues/34681. I will create a separate issue for that.

## Why?

The constrained tabbing feature is broken since more than one year now. It needs to work reliably and consistently across all major browsers. It also needs to be tested more effectively.

## How?

- Adds some actual focus management when a tabbable element contains the next tabbable element. In this case, Safari seems to disagree with other browsers when there is no focused element and the 'Sequential focus navigation starting point' feature comes into play.
- Determines whether the modal dialog content is scrollable, that is: when there is an actual content overflow.
- Makes the scrollable div focusable, only when there is an actual content overflow. This reproduces the native Firefox behavior for all browsers.
- Use `ResizeObserver` to listen to the modal dialog content size changes and checks again whether it's scrollable. This is necessary because:
  - When resizing browser window, zooming in/out, changing orientation on mobile devices, etc., the content may overflow or not depending on the actual size of the content area.
  - Sometimes the modal dialog content is dynamic so that its size changes on the fly, e.g. the Preferences modal uses tabs. Some tabs produce a content overflow, some don't.
- Wraps the dialog content within a `<div>` element: this is necessary to determine size changes via `ResizeObserver`.
- Adds `aria-label="Scrollable section" tabindex="0"` to the dialog content when it's scrollable 
- Adds a `scrollableContentLabel` prop that can be used to override the default label `Scrollable section`.
- Adds a basic focus style to the content when it's scrollable and focusable.
- Makes the a11y E2E test run also with Firefox and Webkit.
- Adds E2E test for the scrollable section.

Note:
- I opted to not use `useResizeObserver` for now. I'm not sure I like the current implementation which injects an additional `<div>` element and overlays it on the node to check for resizing. Also, that works only when the node to check for is a positioned element. 
- It would be great to abstract and futher improve the scrollable section functionality in https://github.com/WordPress/gutenberg/issues/45809. Before that, it would be great to wait for https://github.com/WordPress/gutenberg/pull/41001 which aims to refactor `useResizeObserver` so that it can be used directly on an element. /Cc @youknowriad 

## Testing Instructions




- Run `npm run test:e2e:playwright -- --headed editor/various/a11y.spec.js` and check the tests pass.
- Edit a post and open the Keyboard shortcuts modal dialog.
- Tab through the elements within the modal dialog and check the list of shortcuts container receives focus and shows a focus style.
- With the scrollable element focused, check you can scroll the element by using the Down and Up arrow keys or the Spacebar / Shift+Spacebar. With Safari you may need to press Fn+Down/Up arrow.
- Check you can cycle through the Close button and the scrollable element by pressing Tab and Shift+Tab.
- Close the Keyboard shortcuts modal dialog and open the Preferences one.
- Depending also on your screen size, the General tab should not produce content overflow so it should not be scrollable.
- Inspect the source and check the content area, that is the element with `role=document` does _not_ have an `aria-label` attribute nor a `tabindex` attribute.
- Click on the Blocks tab. This tab has longer content and, depending also on your screen size, it should be scrollable.
- Check the content area is now scrollable and does have `aria-label="Scrollable section" tabindex="0"` attributes.
- Type `qwerty` in the 'Search for a block' search field.
- The modal dialog content will be shorter and only display 'No blocks found'.
- Check the content area is no longer scrollable nor focusable.
- Check it does _not_ have an `aria-label` attribute nor a `tabindex` attribute.
- Clear the search field. The content will show the long list of blocks again.
- Check the content area is scrollable and focusable again.
- Check it does have `aria-label="Scrollable section" tabindex="0"` attributes.
- Check there are no visual regressions in other modal dialogs e.g.:
  - the 'Explore all patterns' modal dialog
  - the confirm dialog when switching back to a draft
  - the confirm dialog when changing post visibility to 'Private'
  - the 'Lock block' modal dialog
  - the Query loop block > 'Choose a pattern' modal dialog
  - any other complex modal dialog e.g. in the Site Editor


### Testing Instructions for Keyboard

See above.

## Screenshots or screencast
